src/main/java/com/iexec/worker/feign/LoginService.java

```diff
@@ -48,48 +47,42 @@ public String getToken() {
         return jwtToken;
     }
 
-    @Override
     public String login() {
+        final String oldToken = jwtToken;
         expireToken();
 
         String workerAddress = credentialsService.getCredentials().getAddress();
         ECKeyPair ecKeyPair = credentialsService.getCredentials().getEcKeyPair();
 
-        String challenge = getLoginChallenge(workerAddress);
+        ResponseEntity<String> challengeResponse = coreClient.getChallenge(workerAddress);
+        if (!challengeResponse.getStatusCode().is2xxSuccessful()) {
+            log.error("Cannot login, failed to get challenge [status:{}]", challengeResponse.getStatusCode());
+            return "";
+        }
+        String challenge = challengeResponse.getBody();
         if (StringUtils.isEmpty(challenge)) {
-            log.error("Cannot login since challenge is empty [challenge:{}]", challenge);
+            log.error("Cannot login, challenge is empty [challenge:{}]", challenge);
             return "";
         }
 
         Signature signature = SignatureUtils.hashAndSign(challenge, workerAddress, ecKeyPair);
-        String token = requestLogin(workerAddress, signature);
+        ResponseEntity<String> tokenResponse = coreClient.login(workerAddress, signature);
+        if (!tokenResponse.getStatusCode().is2xxSuccessful()) {
+            log.error("Cannot login, failed to get token [status:{}]", tokenResponse.getStatusCode());
+            return "";
+        }
+        String token = tokenResponse.getBody();
         if (StringUtils.isEmpty(token)) {
-            log.error("Cannot login since token is empty [token:{}]", token);
+            log.error("Cannot login, token is empty [token:{}]", token);
             return "";
         }
 
         jwtToken = TOKEN_PREFIX + token;
+        log.info("Retrieved {} JWT token from scheduler", Objects.equals(oldToken, jwtToken) ? "existing" : "new");
         return jwtToken;
     }
 
-    private void expireToken() {
+    public void expireToken() {
```

Comment:
private?

Reply:
I guess we want it public so any service can tell the login service the token is expired when it receives a 4xx HTTP error.
But we may do it with events. This could be a better pattern.

Reply:
I went back to `private`. There was no reason to modify this here.
See c76e9f5